pkg/util/testing/client.go

```diff
@@ -18,22 +18,42 @@ package testing
 
 import (
 	"context"
+	"errors"
 	"fmt"
 	"strings"
 	"sync"
 
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/types"
 	clientgoscheme "k8s.io/client-go/kubernetes/scheme"
 	"sigs.k8s.io/controller-runtime/pkg/client"
 	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/client/interceptor"
 
 	kueue "sigs.k8s.io/kueue/apis/kueue/v1beta1"
 	"sigs.k8s.io/kueue/pkg/controller/core/indexer"
 )
 
 func NewFakeClient(objs ...client.Object) client.Client {
-	return NewClientBuilder().WithObjects(objs...).WithStatusSubresource(objs...).Build()
+	return NewClientBuilder().WithObjects(objs...).WithStatusSubresource(objs...).
+		WithInterceptorFuncs(interceptor.Funcs{SubResourcePatch: func(ctx context.Context, clnt client.Client, SubResourceName string, obj client.Object, patch client.Patch, opts ...client.SubResourcePatchOption) error {
+			// Apply patches are supposed to upsert, but fake client fails if the object doesn't exist,
+			// if an apply patch occurs for an object that doesn't yet exist, create it.
+			if patch.Type() != types.ApplyPatchType {
```

Comment:
Using a debugger, I can confirm that we are hitting this function now.

I don't really know what behavior we actually want in this function though. I took this code from an example about how to work around SSA patch.

Reply:
The existing logic is just trying to use a Create instead of a Patch when the object doesn't exist. That's one part of the problem, although we never use SSA for objects that don't exist in Kueue. The next one is the patch type. We need to make it a StrategicMergePatch.